llvm/docs/LangRef.rst

```diff
@@ -1185,6 +1189,9 @@ Currently, only the following parameter attributes are defined:
     value should be sign-extended to the extent required by the target's
     ABI (which is usually 32-bits) by the caller (for a parameter) or
     the callee (for a return value).
+``noext`` This indicates to the code generator that the parameter or return
+    value has the high bits undefined, as for a struct in register, and
+    therefore does not need to be sign or zero extended.
```

Comment:
I don't understand how this is different from the default behavior already

Reply:
It's not.

The issue is that it's easy to write LLVM code that accidentally skips ABI computations.  "noext" is an explicit marker indicating that the code generating the IR does understand the ABI, and decided no extension is necessary.  That lets the backend detect whether some transform skipped doing ABI computations.

Reply:
Can we do IR verifier for ABI illegal cases instead?

Reply:
I don't understand how this improves the situation, or how the IR can "understand the ABI". The IR has its own ABI. This may or may not match the C ABI, the IR ABI is not really defined in terms of C

Reply:
There doesn't exist an "IR ABI". If you don't specify a calling convention in IR, [it must be C calling convention](https://llvm.org/docs/LangRef.html#calling-conventions).
Or you want to call representations out of C ABI definition "IR ABI"? I cannot agree with either. A typical ABI has two merits: well documented and stable. Neither do they have. They are almost exceptions unconsciously generated front end or written by hand. They exist just because we lack a verifier for them.
My suggestion is to introduce ABI verification code for each target in the IR verifier like [this](https://github.com/llvm/llvm-project/blob/main/llvm/lib/IR/Verifier.cpp#L2333-L2339). We may start from cases we care most and gradually check all rules followed by clang front end.

Reply:
"Skipped ABI computations", meaning code that assumes that C int(int) can be blindly translated to i32(i32).  Like https://reviews.llvm.org/D133949.  The goal is to get some sort of error from code like that, instead of an obscure miscompile.

Reply:
> Can we do IR verifier for ABI illegal cases instead?

In previous discussions it has been agreed to generally let the target decide on how to act on the argument extensions attributes. So we decided to keep the check for these in the (SystemZ) backend. Per my understanding, this is related to the target platform ABI - some targets like SystemZ has specific extension requirements of arguments. Others do not and can therefore ignore these attributes. Therefore, having a general IR verifier for this doesn't quite make sense - unless one would add a new hook as well to check with the target that it depends on valid argument extension attributes.

> My suggestion is to introduce ABI verification code for each target in the IR verifier...

I agree this would make sense, but think this could be a follow-up patch once we have the needed parts this patch provides.

> The goal is to get some sort of error from code like that, instead of an obscure miscompile.

Yes, if the front-end or an instrumentation pass forgets to add the extension, it could lead to wrong-code that is hard to track down. This is especially treacherous since somebody might work on a target that do not care about this at all and introduce a problem on another target that depends on it. In such a case it would be great if e.g. the SystemZ buildbot would fail by detecting the case with no extension attribute.

Reply:
> There doesn't exist an "IR ABI".

There absolutely is. A given IR signature has a backend interpretation. That is the IR ABI. It doesn't matter if this is documented or standardized in any way, it is what it is.

> My suggestion is to introduce ABI verification code 

LLVM is not a clang / C backend

Reply:
- That's incorrect. Some implementation defined behavior are subject to change, hence cannot be called an ABI. It just like we cannot assume UB code get the same result between different compilers or from version to version.
- If other front end doesn't specific a calling convention, then it uses C calling convention. See my previous link.

BTW, I'm not opposite to this patch.

Reply:
> * That's incorrect. Some implementation defined behavior are subject to change, hence cannot be called an ABI. It just like we cannot assume UB code get the same result between different compilers or from version to version.
> * If other front end doesn't specific a calling convention, then it uses C calling convention. See my previous link.

Sorry, I don't follow - what part is incorrect according to you? The *target* ABI is independent of the implementation - it is how things are done on a specific platform. On some machines this type of extensions *must* be done, and this will not change. Other machines do not care about this at all. How could this become more clear in the LangRef.rst do you mean?

Reply:
> Sorry, I don't follow - what part is incorrect according to you?

Oh, I should have quoted what I replied for
>> There doesn't exist an "IR ABI".
>
> There absolutely is. A given IR signature has a backend interpretation. That is the IR ABI. It doesn't matter if this is documented or standardized in any way, it is what it is.

For the question

> The target ABI is independent of the implementation - it is how things are done on a specific platform. On some machines this type of extensions must be done, and this will not change. Other machines do not care about this at all. How could this become more clear in the LangRef.rst do you mean?

The target ABI (from C's perspective) is a subset of what IR can represent. I mean to those representation out of the scope. They are subject to change, because ABI can be defined rather late than the implementation and possible different with it. An example is the `half` type on X86. It uses GPR for a couple of year by the implementation before ABI specifies FPR.